Fix various ISIS SANS clipboard bugs

DavidFair opened the pull request:
**Description of work.**
- Fixes all the edge cases found in #33195 
- Port (copy and paste) the existing implementation into the table model and surround it in tests

See commits for more details

**To test:**
- Download the ISIS Sample Data
- Locate the `loqdemo` folder in the sample data
- Open the ISIS SANS interface
- Click batch file, load the CSV
- Click the row with name `first_row` and shift click select down to `second_row`, click copy
- Click off the rows (so none is selected), click paste and check the order copied correctly
- Do the same steps again, but this time selecting `second_row` first, then `first_row`
- Check the pasted order is goes first->second again
- Try breaking copy / cut paste various different ways 

Fixes #33286 

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Ensure RowEntries __eq__ works correctly

Fixes RowEntries __eq__ as this was not working correctly and our unit
tests for table model will implicitly rely on them.
Re #33286

DavidFair committed:
Port/Fix copy-paste handling into table model

Copy and pastes the table handling from the presenter (which is
incorrect) into the table model which should handle this all.

Add unit tests reflecting every case the testing team came up to find
various edge cases and fix them in the implementation.
Re #33286

DavidFair committed:
Move impl from presenter to model

Moves the implementation from the presenter to the model, massivly
simplifying testing and the (already too large) presenter.

This Fixes #33286

DavidFair committed:
Adding missing __hash__ override

Since we have overridden _eq_ we need to also handle hash.

This is a bit different, a lot of our code assumes that two instances of
A will duplicate into a dictionary. So we must preserve the object impl,
taking the memory address as the hash value.

RichardWaiteSTFC reviewed (changes requested):
Looks like another edge case is needed when pasting multiple rows to an empty row when only trailing empty rows exist below?
![image](https://user-images.githubusercontent.com/55979119/149375103-accf4f92-e8db-4ef4-afe3-f8125e50e09b.png)

DavidFair committed:
Rework UX for copy-paste

The previous flow for copy and paste was too clever, and tried to be
safe. Having tested other spreadsheet tools the best thing to do is the
obvious thing (even if a user can get it wrong).
That way it always feels intuative

Re #33286

DavidFair commented:
On discussion me and Richard agreed to push this to 6.4, as it borderlines a new feature
The current feature (whilst clunky) does not cause crashes / problems so we agreed to play it safe

RichardWaiteSTFC reviewed (dismissed):
This looks good, behaves like other spreadsheet software which I find quite intuitive. Test coverage looks good for many edge cases.

martyngigg commented on docs/source/release/v6.3.0/33286_sans.rst:
```diff
@@ -0,0 +1,4 @@
+Bugfixes
```
If we're targeting this at 6.4 can we move the release notes to the 6.4 section?

DavidFair committed:
Move release notes from 6.3 -> 6.4

Since we decided the risk/reward for this change and the number of users
running into it favours pushing this into the next release

RichardWaiteSTFC commented:
@DavidFair - sorry forgot about this. Do I need to test it again, or did you just change the release notes in your force push?

DavidFair commented:
It was a rebase from release-next to main so I could shift the release notes into 6.4
This should be identical code-wise to before

RichardWaiteSTFC commented:
> @DavidFair - just thought I'd try this for peace of mind. Looks like it's functioning correctly but is very slow, it hangs for ~5s when pasting. Probably this is my computer and there's no reason your code is slowing it down, I was wondering whether you could check?

Ah the slow down appears to only occur when I have my default instrument set as WISH as opposed to LOQ. I seem to recall this produced an error, I no longer see this but the slow down is reproducible.

DavidFair commented:
> > @DavidFair - just thought I'd try this for peace of mind. Looks like it's functioning correctly but is very slow, it hangs for ~5s when pasting. Probably this is my computer and there's no reason your code is slowing it down, I was wondering whether you could check?
> 
> Ah the slow down appears to only occur when I have my default instrument set as WISH as opposed to LOQ. I seem to recall this produced an error, I no longer see this but the slow down is reproducible.

This is because the table opens the associated .nxs file to get the instrument name. Unfortunately it does the `n` times for `n` rows because of a long-standing bug (I can't find the issue # at the moment)

RichardWaiteSTFC reviewed (approved):
> > Ah the slow down appears to only occur when I have my default instrument set as WISH as opposed to LOQ. I seem to recall this produced an error, I no longer see this but the slow down is reproducible.
> 
> This is because the table opens the associated .nxs file to get the instrument name. Unfortunately it does the `n` times for `n` rows because of a long-standing bug (I can't find the issue # at the moment)

OK I'll approve!
